Monitor: Tags with metadata

chakflying opened the pull request:
## Add tags to monitors, with optional metadata for each tag.

In the create / edit monitor page, the user can add new or select existing tags to apply to the current monitor. Any changes is stored as draft data in the `tagsManager` component, and will only be committed when the user click `Save` in the parent component. This made the code a bit more complicated, but I think the results are worth it.

## Implementation

Since the `Add tag` -> `Add monitor_tag` operation is order dependent, I wrapped the `socket.io` call in a Promise to be able to `await` it. I find this pattern way more usable then the default simple callback.

With a simple UX being top priority, I think there should be no need for a separate UI for managing tags. Instead, every time a `deleteMonitorTag` is triggered, it will remove any `tags` which is unused. So there is currently no way to change the color or name of an existing tag. Let me know if you think this should be added.

Similarly, although the DB supports any hex color for the tag, the user can only select a predefined set of colors for now, since putting in a color picker is more complicated, and if we allow any arbitrary color, we may need to change the tag font-color depending on the brightness of the selected background, which I don't know how except maybe a css filter.

I also realized I made a mistake before by putting multiselect styling in the `editMonitor` component. At this point maybe it should be split into a separate file instead. But I think we can merge this first and refactor later.

|light|dark|
|----|----|
|![image](https://user-images.githubusercontent.com/3271800/132821626-9f45ee3d-28e4-44f9-b5d6-84509fe3af41.png)|![image](https://user-images.githubusercontent.com/3271800/132821717-d6f59d3b-bed8-41c0-a94f-5a55c8012dab.png)|


|light|dark|
|----|----|
|![image](https://user-images.githubusercontent.com/3271800/132169382-a5641594-242f-4148-bc6c-e67485dbd742.png)|![image](https://user-images.githubusercontent.com/3271800/132157880-0144b3e6-f84b-42c3-9b25-3030913d03eb.png)|

Edit Monitor:
|light|dark|
|----|----|
|![image](https://user-images.githubusercontent.com/3271800/132821895-97dd17d5-958d-48c4-a73f-c56344c2c76c.png)|![image](https://user-images.githubusercontent.com/3271800/132821946-488b14c1-a827-4403-b670-bf91c4462547.png)|




- [x] [Database] Create tags table and monitor_tags table
- [x] [Backend] Handle CRUD of tags
- [x] [Backend] Handle assign, update and remove of tags
- [x] [Frontend] Monitor Details: display tags
- [x] [Frontend] Monitor Edit: display selected tags, form to add existing or new tags, remove selected tags
- [x] [Frontend] Monitor List: display tags

Optional:
- [x] [Frontend] Monitor List: input to filter by tags

chakflying commented:
@louislam Core functionality should be done now, feel free to review or make any styling changes. 

For the filter, I'm thinking if it should be a universal <input> field which works for both the monitor name and tags, or just a dropdown where user can select tags. Seems like a input field is better, but I'll need to think about the implementation.

Ponkhy commented:
Looks great! I really like that feature!
Should keep in mind to update the database version number and file name when it gets merged into the master branch.
Maybe also give the user the option on mobile to hide the tags in the monitor list.

louislam commented:
Just tested, really a nice feature!

Not sure whether it is hard or not. If a selected tag could be added without clicking the "add" button, it would be a little better.  

![image](https://user-images.githubusercontent.com/1336778/132558393-74ba7201-83d6-44d8-8d38-9b5f99924e4b.png)


> So there is currently no way to change the color or name of an existing tag. Let me know if you think this should be added.

That is ok, nothing can be perfect in the first place.

chakflying commented:
> If a selected tag could be added without clicking the "add" button,

@louislam But then there is no way for the user to enter the optional value?

louislam commented:
> > If a selected tag could be added without clicking the "add" button,
> 
> @louislam But then there is no way for the user to enter the optional value?

Oh, I though "value" belongs to the tag.

chakflying commented:
The value is stored in the `monitor_tag` relations table, so theoretically the user can add duplicate tags with different values.

![image](https://user-images.githubusercontent.com/3271800/132629122-f915fed5-57cb-441e-8b53-4d63061ba72b.png)

louislam commented:
Do you agree if I change this part into a dialog? 

As I personally just use the tag without value, I thought the dropdown list is similar to the http status codes list and forget to click "Add".

Split this part into a dialog, I think users would not miss it, as the users need to click "Add" or "X" either.

![image](https://user-images.githubusercontent.com/1336778/132670349-74a46fef-c43f-4baf-85c6-5fda60a80aee.png)

chakflying commented:
Sure! That sounds like a good idea. I thought of implementing it with the taggable multi-select, unfortunately I can't think of a way to include the logic of choosing colors and adding value.

louislam commented:
Found two issues:
1.mp4: Remove a same tag by `x`, all tags with same tag name but different values also be removed.
2.mp4: Not quite sure how I triggered it, but sometimes, it cannot be add a if I removed previously.  

Testing version: https://github.com/chakflying/uptime-kuma/commit/64d67223fc0dbd3181760ff1560f496ccf824d33

https://user-images.githubusercontent.com/1336778/132703678-bc51f9e6-6e7e-459d-8d8e-48168006cb99.mp4

https://user-images.githubusercontent.com/1336778/132703690-81bea801-08c4-4c95-a3bb-4d2497ab175e.mp4

chakflying commented:
First problem should be fixed. Apparently boolean logic is hard, and `!a and !b != !(a and b)`. 🙄

louislam commented:
> First problem should be fixed. Apparently boolean logic is hard, and `!a and !b != !(a and b)`. 🙄

Lol, of course, `!(a && b)` after substitution is `!a || !b`

louislam commented:
Found more issues:

1. Some warnings in Chrome console
2. `3.mp4`: Cannot add another tag with same name but different value (e.g. added DNS, want to also add DNS:TypeA)
3. `4.mp4`: After reload the edit page, I can add DNS:TypeA now. Save, and then remove this tag. Save again, the save button is broken.

![image](https://user-images.githubusercontent.com/1336778/132804858-7b693700-3409-437e-be10-e93aaee163a8.png)


https://user-images.githubusercontent.com/1336778/132807975-90d8c4b9-0238-45b8-aef3-328420156ad8.mp4


https://user-images.githubusercontent.com/1336778/132807981-bb1f93f0-fcd3-4fb3-a80f-757e79ae50a1.mp4

chakflying commented:
FYI I don't like to "merge from master to {feature branch}" since it adds an unnecessary commit and makes the final tree more complicated. I regularly rebased the feature branch onto master instead.

louislam commented:
> FYI I don't like to "merge from master to {feature branch}" since it adds an unnecessary commit and makes the final tree more complicated. I regularly rebased the feature branch onto master instead.

Sorry about that.

And I find that this feature is bigger than I thought. I may need more time to test it. And as previously mentioned, I want to put a part into a dialog, so it would be delayed to 1.7.0.

Free feel to mention me again when it is ready to test again. Thanks for contributing.👍

chakflying commented:
@louislam previous problems should be fixed. Rebased on master, split adding tags to a modal, added special handling for re-adding a removed tag & reusing the same tag before saving. Sorry for the trouble!

chakflying commented:
rebased on master to resolve conflicts.

louislam commented:
It is much more stable now!

Just found a small issue. The form is submitted unexpectedly while pressing enter here:

![image](https://user-images.githubusercontent.com/1336778/133130825-dfd034fe-f50a-4711-9f6b-7c1ef0ab6ca8.png)

chakflying commented:
Since the modal is still technically in the same form as the EditMonitor, I don't know how to block the browser default submit behaviour (`@submit.stop` doesn't seem to work). Instead I just added a function to block the enter key if the add tag form is invalid.
